### PR TITLE
#171380746 User descriptive signup validation errors

### PR DESCRIPTION
**Pull request by muchristian**
**What does this PR do?**
for descriptive user signup validation errors

**Description of Task to be completed?**
add src/middlewares/authValidation.js middleware
add src/middlewares/validation.js service
modify and add if email exist script in src/controllers/authentication.controller.js
add test script in src/controllers/authentication.test.js

**How should this be manually tested?**
- clone the project
- open your project in cmd, and then write _npm install_ to install packages
- while signing up try to write invalid text on email for instance.
- run the project by writing in cmd _npm start_ . 
- test signup endpoint in postman for instance, and see the output. Actually will return error because email is invalid as aforementioned.

**Any background context you want to provide?**
N/A

**What are the relevant pivotal tracker stories?**
[#171380746](https://www.pivotaltracker.com/n/projects/2435482/stories/171380746)

### CODE REVIEW

**Review by ricahcyuzuzo — CHANGES_REQUESTED** [review]
Would you solve the coverage issue?

File: src/controllers/authentication.controller.js [hunk 1]
```diff
@@ -26,6 +26,10 @@ export default class AuthenticationController {
   static async signUp(req, res) {
     try {
       const payload = await getFormData(req.body);
+      const checkEmail = await findUserByEmail(payload.email);
+      if (checkEmail) {
+        return errorResponse(res, 409, 'user with the provided email exist');
```
Comment (Muhire-Josue) [inline src/controllers/authentication.controller.js]: can you use the `../utils/customMessages.js` and `../utils/statusCodes.js`. You can just move the status code and error message there and call it here.

File: src/middlewares/authValidation.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * @param {object} res
+ * @param {object} next
+ * @returns{object} return next() if validations pass
+ */
+const signupValidation = async (req, res, next) => {
+    const { error } = validateSignup(req.body);
+    if (error) {
+        const { details } = error;
+    const messages = details.map((err) => err.message.replace(/['"]/g, '')).join(', ');
+    return res.status(400).json({ status: 400, error: messages });
```
Comment (Muhire-Josue) [inline src/middlewares/authValidation.js]: same thing here, you can move that status code in `../utils/statusCodes.js` then call it here.

File: tests/authentication.test.js [hunk 3]
```diff
@@ -5,13 +5,23 @@ import customMessages from '../src/utils/customMessages';
 import statusCodes from '../src/utils/statusCodes';
 import mockData from './data/mockData';
 
-const { signupData, incompleteData } = mockData;
+const { signupData, invalidData, signupData1 } = mockData;
 
 chai.use(chaiHttp);
 chai.should();
 
 // Signup
```
Comment (Muhire-Josue) [inline tests/authentication.test.js]: could you remove this comment please?

File: tests/authentication.test.js [hunk 4]
```diff
@@ -27,16 +37,13 @@ describe('User sign up', () => {
...
       .request(server)
       .post('/api/auth/signup')
-      .send(incompleteData)
+      .send(signupData)
       .end((err, res) => {
-        const { error } = res.body;
-        expect(res.status).to.equal(statusCodes.badRequest);
-        expect(error);
-        expect(error).to.equal(customMessages.userSignupFailed);
+        expect(res.status).to.equal(409);
```
Comment (Muhire-Josue) [inline tests/authentication.test.js]: Also you can use the status from `../utils/statusCodes.js` here.

**Review by Muhire-Josue — COMMENTED** [review]
You did a great job here, congratulations. We have predefined response messages and status code in `../utils/customMessage.js` and `../utils/statusCodes.js` could you refer to those in every response and in the tests. 
                 `ex: successResponse(res, message.userCreated, statusCode.ok)`
                  `ex: expect(res.body.status).to.equal(statusCode.conflict)`

**Review by kipropbrian — APPROVED** [review]

File: tests/authentication.test.js [hunk 5]
```diff
@@ -5,13 +5,23 @@ import customMessages from '../src/utils/customMessages';
...
-// Signup
+
 describe('User sign up', () => {
+  it('Should return 400', (done) => {
+    chai
+      .request(server)
+      .post('/api/auth/signup')
+      .send(invalidData)
+      .end((err, res) => {
+        expect(res.status).to.equal(statusCodes.badRequest);
```
Comment (the22mastermind) [inline tests/authentication.test.js:32]: Could you also expect an error message and pass the appropriate message?

File: tests/authentication.test.js [hunk 6]
```diff
@@ -27,24 +37,18 @@ describe('User sign up', () => {
...
       .request(server)
       .post('/api/auth/signup')
-      .send(incompleteData)
+      .send(signupData)
       .end((err, res) => {
-        const { error } = res.body;
-        expect(res.status).to.equal(statusCodes.badRequest);
-        expect(error);
-        expect(error).to.equal(customMessages.userSignupFailed);
+        expect(res.status).to.equal(statusCodes.conflict);
```
Comment (the22mastermind) [inline tests/authentication.test.js:138]: Same comment as above

**Review by Ugizwenayo-Divine — COMMENTED** [review]
You really did a good job Christian but I suggested that
- you remove that authentication.js file in src/routes/api as it is empty
- you also remove mocksupdata file within tests folder as Emmanuel said that He didn't use it and he is the one who created it
- Finally remember to fix those hound issues

good work by the way

**Review by Muhire-Josue — APPROVED** [review]
LGTM

**Review by Muhire-Josue — COMMENTED** [review]
Can you take care of those hound violations

**Review by ricahcyuzuzo — APPROVED** [review]
Good Job

File: src/middlewares/validations.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+     return methods;
+};
+
+/**
+* @param {object} user
+* @returns {object} return body assigned to their validation methods
+*/
+const validateSignup = user => {
+    const schema = Joi.object({
+        firstName: validationMethods(/^[a-zA-Z]\S+$/, { 'string.pattern.base': 'firstName must consists of alphabets only and no whitespace' }),
```
Comment (the22mastermind) [inline src/middlewares/validations.js]: Could you move all these custom messages in customMessages.js and import them here?

File: tests/authentication.test.js [hunk 8]
```diff
@@ -5,13 +5,97 @@ import customMessages from '../src/utils/customMessages';
...
+  it('Should return 400 if firstname is invalid', (done) => {
+    chai
+      .request(server)
+      .post('/api/auth/signup')
+      .send(invalidFirstname)
+      .end((err, res) => {
+        const { error } = res.body;
+        expect(res.status).to.equal(statusCodes.badRequest);
+        expect(error);
+        expect(error).to.equal('firstName must consists of alphabets only and no whitespace');
```
Comment (the22mastermind) [inline tests/authentication.test.js]: Could you create these custom messages in customMessages.js and import them here? The same way you used statusCodes.

**Review by the22mastermind — CHANGES_REQUESTED** [review]
Good Job @muchristian check out the comments I left. Keep up the good work.

File: src/utils/customMessages.js [hunk 9]
```diff
@@ -6,7 +6,15 @@ const messages = {
   loginUnauthorized: 'Email and password mismatch',
   loginEmailEmpty: 'Please enter your email',
   loginPasswordEmpty: 'Please enter your password',
-  loginPasswordAndEmailEmpty: 'Please enter your email and your password'
+  alreadyExistEmail: 'user with the provided email exist',
+  loginPasswordAndEmailEmpty: 'Please enter your email and your password',
+  invalidFirstname: 'firstName must consists of alphabets only and no whitespace or special characters',
+  invalidLastname: 'lastName must consists of alphabets only and no whitespace or special characters',
+  invalidUsername: 'userName must consists of alphabets only and no whitespace or special characters',
+  invalidEmail: 'email must consists of alphabets, numbers and @ sign',
```
Comment (mifeille) [inline src/utils/customMessages.js]: Please start all your error messages with capital letters and for the email, you can give an example of how the email should look like (for ex: example@email.com)

File: src/utils/customMessages.js [hunk 10]
```diff
@@ -11,7 +11,15 @@ const messages = {
   notExistUser: 'User does not exist',
   changed: 'The password has been changed',
   notChanged: 'The password not changed',
-  errorMessage: 'occured an error'
+  errorMessage: 'occured an error',
+  alreadyExistEmail: 'user with the provided email exist',
```
Comment (mifeille) [inline src/utils/customMessages.js]: Please change this error to this: The user with the provided email already exists.

File: src/utils/customMessages.js [hunk 11]
```diff
@@ -11,7 +11,15 @@ const messages = {
...
+  invalidFirstname: 'firstName must consists of alphabets only and no whitespace or special characters',
```
Comment (mifeille) [inline src/utils/customMessages.js]: Please change this error to this and do it for the last name.
The first name must consist of alphabets only, no whitespaces or special characters.

**Review by Ugizwenayo-Divine — COMMENTED** [review]
why did you remove run seed from test script
I think this findUserIfEmail exist do the same as findUserByEmail
why did you remove comments that were explaining what a function do?

**Pull request merged**